libcudacxx/include/cuda/std/detail/libcxx/include/__bit_reference

```diff
@@ -1142,56 +1052,79 @@ bool __equal_aligned(
 }
 
 template <class _Cp, bool _IC1, bool _IC2>
-inline _LIBCUDACXX_INLINE_VISIBILITY bool
+inline _LIBCUDACXX_HIDE_FROM_ABI _LIBCUDACXX_INLINE_VISIBILITY _CCCL_CONSTEXPR_CXX14 bool
 equal(__bit_iterator<_Cp, _IC1> __first1, __bit_iterator<_Cp, _IC1> __last1, __bit_iterator<_Cp, _IC2> __first2)
 {
   if (__first1.__ctz_ == __first2.__ctz_)
   {
-    return __equal_aligned(__first1, __last1, __first2);
+    return _CUDA_VSTD::__equal_aligned(__first1, __last1, __first2);
   }
-  return __equal_unaligned(__first1, __last1, __first2);
+  return _CUDA_VSTD::__equal_unaligned(__first1, __last1, __first2);
 }
 
-template <class _Cp, bool _IsConst, typename _Cp::__storage_type>
+template <class _Cp, bool _IsConst>
 class __bit_iterator
 {
 public:
-  typedef typename _Cp::difference_type difference_type;
-  typedef bool value_type;
-  typedef __bit_iterator pointer;
-  typedef typename conditional<_IsConst, __bit_const_reference<_Cp>, __bit_reference<_Cp>>::type reference;
-  typedef random_access_iterator_tag iterator_category;
+  using difference_type = typename _Cp::difference_type;
+  using value_type      = bool;
+  using pointer         = __bit_iterator;
+#ifndef _LIBCUDACXX_ABI_BITSET_VECTOR_BOOL_CONST_SUBSCRIPT_RETURN_BOOL
+  using reference = __conditional_t<_IsConst, __bit_const_reference<_Cp>, __bit_reference<_Cp>>;
+#else
+  using reference = __conditional_t<_IsConst, bool, __bit_reference<_Cp>>;
+#endif
+  using iterator_category = random_access_iterator_tag;
 
 private:
-  typedef typename _Cp::__storage_type __storage_type;
-  typedef typename conditional<_IsConst, typename _Cp::__const_storage_pointer, typename _Cp::__storage_pointer>::type
-    __storage_pointer;
+  using __storage_type = typename _Cp::__storage_type;
+  using __storage_pointer =
+    __conditional_t<_IsConst, typename _Cp::__const_storage_pointer, typename _Cp::__storage_pointer>;
+
   static const unsigned __bits_per_word = _Cp::__bits_per_word;
 
   __storage_pointer __seg_;
   unsigned __ctz_;
 
 public:
-  _LIBCUDACXX_INLINE_VISIBILITY __bit_iterator() noexcept
-#if _CCCL_STD_VER > 2011
+  _LIBCUDACXX_HIDE_FROM_ABI _LIBCUDACXX_INLINE_VISIBILITY _CCCL_CONSTEXPR_CXX14 __bit_iterator() noexcept
+#if _CCCL_STD_VER >= 2014
       : __seg_(nullptr)
       , __ctz_(0)
-#endif
+#endif // C++14+
   {}
-  // avoid re-declaring a copy constructor for the non-const version.
-  using __type_for_copy_to_const = _If<_IsConst, __bit_iterator<_Cp, false>, struct __private_nat>;
 
-  _LIBCUDACXX_INLINE_VISIBILITY __bit_iterator(const __type_for_copy_to_const& __it) noexcept
+  // When _IsConst=false, this is the copy constructor.
+  // It is non-trivial. Making it trivial would break ABI.
+  // When _IsConst=true, this is a converting constructor;
+  // the copy and move constructors are implicitly generated
+  // and trivial.
```

Comment:
We do not have an ABI because we are just bringing this in. Can we just make it trivial?

This is especially important for our use cases where a lot of stuff wants at `trivially_copyable`

Reply:
I was afraid there may be weird consequences of trying to do this, but = default and a sprinkling of sfinae seems to do the job. Done.